04-evm-mapping.md

```diff
@@ -0,0 +1,307 @@
+- Class `3` (BLS key): payload is an inlined BLS public key. Fixed 48 bytes.
+
+In conclusion, the maximum address length in Filecoin is 49 bytes or 392 bits (class 3 address). This creates two problems:
+
+1. The worst case scenario is larger than the width of the Ethereum address type. Even if BLS addresses were prohibited in combination with EVM actors, class 1 and class 2 still miss the limit by 1 byte (due to the prefix).
+2. It exceeds the EVM's 256 bit architecture.
+
+Problem 1 renders Solidity smart contracts instantly incompatible with the Filecoin addressing scheme, as well as EVM opcodes that take or return addresses for arguments, e.g. CALLER, CALL, CALLCODE, DELEGATECALL, COINBASE, etc. This problem is hard to work around, and would require a fork of the EVM to modify existing opcodes for semantic awareness of addresses (although this is really hard to get right), or to introduce a Filecoin-specific opcode family to deal Filecoin addresses (e.g. FCALL, FCALLCODE, etc.) The latter would break as-is deployability of existing smart contracts.
+
+Problem 2 can be workable by spilling over and combining values in the stack, through Filecoin-specific Solidity libraries.
+
+Find the proposed solution next, and a summary of alternatives considered in Annex A.
+
+### Proposed solution: universal stable addresses
+
+The idea is to promote address class 2 into a unified fixed-size stable address space all actors, **including account actors**. At present, this class only identifies **non-account actors**. The only address space that covers all actors is the ID class, which is not reorg-stable and thus unsafe.
+
+The width of this address space is already 20 bytes. EVM actors would only interact with class 2 addresses, with the prefix stripped to honour the 160-bit address expectation, for full EVM bytecode compatibility.
+
+Nevertheless, it has been noted that the security of 20-byte addreses is beginning to be insufficient, with an estimated cost of $10bn to find an arbitrary collision by carrying out 2^80 computations.
+
+Therefore, the actual proposal is **to introduce a new address class 4 for a 256-bit stable address space**. This class would become the **canonical stable address class**, with class 2 addresses (20-byte) mapping over to their class 4 equivalents (32 bytes). Class 2 can thus be thought of as an alias/symlink to class 4, to be used in environments that are unable handle larger widths.
+
+The state tree would track class 2 and class 4 address for all actors by maintaining the necessary indices.
+
+**Class 2 and 4 address derivation**
+
+Class 2 and 4 addresses are protocol-derived, by hashing the relevant cryptographic or input material.
+
+With a high probability, the hash function of class 4 will remain BLAKE2, with a 256-bit digest size. Class 2 will continue relying on blake2b-160.
```

Comment:
So... we could save some space if we define class 2 as a truncation of class 4. Instead of mapping both class-2 and class-4 addresses to ID addresses, we'd map class 2 addresses to (ID, class4-suffix) tuples.

We'd want to determine how much space this map is really taking up on chain before making any decisions (this optimization may not be worth it).